### PR TITLE
as a user, i do not want jslint to crash from excessive-recursion

### PR DESCRIPTION
**Pull request by kaizhu256**
sometimes in nodejs, when async/simultaneous linting multiple-files with large number of _[machine-generated]_ leading-newlines, jslint will crash with a ```RangeError("Maximum call stack size exceeded")```

this patch mitigates the above scenario by avoiding unnecessary-recursion while parsing leading-emptylines.

live web-demo of patch available at https://kaizhu256.github.io/JSLint/branch.remove_unnecessary_recursion/

1. screenshot - jslint crashing from recursing through excessive leading-emptylines
screen shot 2018-11-04 at 6 09 25 pm

2. screenshot - patched jslint no longer crashes
screen shot 2018-11-04 at 6 11 42 pm

### CODE REVIEW

**Comment by douglascrockford** [comment]
The ES6 Standard requires that engines not produce that error. Use a standards compliant JavaScript engine.

**Comment by kaizhu256** [comment]
@douglascrockford, your suggestion is impractical on the server-side, where v8/nodejs is the only javascript-engine available for most developers.

you may have a point that v8 shares the blame for refusing to implement PTC, but having this issue remain in jslint for indefinite future, makes jslint unsuitable for production-use, as it causes random/flaky crashes in nodejs' nondeterministic gc (i'm constantly guessing whether or not my server will crash while its auto-jslinting files in background).

**Comment by douglascrockford** [comment]
File bug reports with Node and V8.

**Comment by kaizhu256** [comment]
i did file a related v8-bug at https://bugs.chromium.org/p/v8/issues/detail?id=8234 (filing a nodejs bug is pointless, as its a v8 problem).

the v8-team assigned it a priority 3, and i don't see them in a hurry to fix it (if ever).  iirc, they actually removed their ptc-code, as it was bit-rotting.  and tc39 is indifferent/in-limbo about engines not following the spec; v8-team made strong performance-related arguments about their decision to revert-implementation [1].

in short,  the only way this issue will get fixed in forseeable-future is by us alone.

[1] tc39 2016-05-25 meeting with no consensus on implementing/enforcing es6 ptc-spec.
https://github.com/rwaldron/tc39-notes/commit/8535bca63c760a43c9d9e0e833dbebf6aeb349c8#diff-f5913216f5ca569e27c22de2b2ecef31R1019

**Comment by kaizhu256** [comment]
can we revisit this pull-request to un-recurse `function lex()`?

1. proper-tail-call spec seems to have no future in any of the js-implementations besides jsc

2. i've been using this patch on my custom jslint with no problems since this pull-request was submitted:

```js
/*
 * patched jslint i've been using since november 2018 at
 * https://github.com/kaizhu256/node-utility2/blob/2019.8.1/lib.jslint.js#L3215
 */
-        if (!source_line) {
-            source_line = next_line();
-            from = 0;
-            return (
-                source_line === undefined
-                ? (
-                    mega_mode
-                    ? stop_at("unclosed_mega", mega_line, mega_from)
-                    : make("(end)")
-                )
-                : lex()
-            );
-        }
+        // jslint-hack - un-recurse
+        while (!source_line) {
+            source_line = next_line();
+            from = 0;
+            if (source_line === undefined) {
+                return (
+                    mega_mode
+                    ? stop_at("unclosed_mega", mega_line, mega_from)
+                    : make("(end)")
+                );
+            }
+        }
```

**Comment by aonghusonia** [comment]
I've been using the babel plugin for tail call optimization with success.  Maybe you could run jslint through back with only that plugin.

https://github.com/krzkaczor/babel-plugin-tailcall-optimization/issues?q=is%3Aissue+is%3Aclosed

**Comment by EvgenyOrekhov** [comment]
@aonghusonia This was fixed in https://github.com/douglascrockford/JSLint/commit/efefb7d4e22359b6fb1977d33712bcc2fda95f14.
